web/src/hotkey.js

```diff
@@ -157,81 +195,129 @@ const keydown_either_mappings = {
-    85: {name: "mark_unread", message_view_only: true}, // 'U'
-    86: {name: "view_selected_stream", message_view_only: false}, // 'V'
-    // The shortcut "a" dates from when this was called "All messages".
-    97: {name: "open_combined_feed", message_view_only: true}, // 'a'
-    99: {name: "compose", message_view_only: true}, // 'c'
-    100: {name: "open_drafts", message_view_only: true}, // 'd'
-    101: {name: "edit_message", message_view_only: true}, // 'e'
-    103: {name: "gear_menu", message_view_only: true}, // 'g'
-    104: {name: "vim_left", message_view_only: true}, // 'h'
-    105: {name: "message_actions", message_view_only: true}, // 'i'
-    106: {name: "vim_down", message_view_only: true}, // 'j'
-    107: {name: "vim_up", message_view_only: true}, // 'k'
-    108: {name: "vim_right", message_view_only: true}, // 'l'
-    109: {name: "move_message", message_view_only: true}, // 'm'
-    110: {name: "n_key", message_view_only: false}, // 'n'
-    112: {name: "p_key", message_view_only: false}, // 'p'
-    113: {name: "query_streams", message_view_only: true}, // 'q'
-    114: {name: "reply_message", message_view_only: true}, // 'r'
-    115: {name: "toggle_conversation_view", message_view_only: true}, // 's'
-    116: {name: "open_recent_view", message_view_only: true}, // 't'
-    117: {name: "toggle_sender_info", message_view_only: true}, // 'u'
-    118: {name: "show_lightbox", message_view_only: true}, // 'v'
-    119: {name: "query_users", message_view_only: true}, // 'w'
-    120: {name: "compose_private_message", message_view_only: true}, // 'x'
-    122: {name: "zoom_to_message_near", message_view_only: true}, // 'z'
+const character_mappings = {
+    "*": {name: "open_starred_message_view", message_view_only: true},
+    "+": {name: "thumbs_up_emoji", message_view_only: true},
+    "=": {name: "upvote_first_emoji", message_view_only: true},
+    "-": {name: "toggle_message_collapse", message_view_only: true},
+    "/": {name: "search", message_view_only: false},
+    ":": {name: "toggle_reactions_popover", message_view_only: true},
+    "<": {name: "compose_forward_message", message_view_only: true},
+    ">": {name: "compose_quote_message", message_view_only: true},
+    "?": {name: "show_shortcuts", message_view_only: false},
+    "@": {name: "compose_reply_with_mention", message_view_only: true},
+};
+
+function is_printable_ascii(key) {
+    return key.length === 1 && key >= " " && key <= "~";
+}
+
+const NAMED_KEY_ATTRIBUTE_VALUES = new Set([
+    "Alt",
+    "ArrowDown",
+    "ArrowLeft",
+    "ArrowRight",
+    "ArrowUp",
+    "Backspace",
+    "Control",
+    "Delete",
+    "End",
+    "Enter",
+    "Escape",
+    "Home",
+    "Meta",
+    "PageDown",
+    "PageUp",
+    "Shift",
+    "Tab",
+]);
+
+const CODE_TO_QWERTY_CHAR = {
+    BracketLeft: "[",
+    Equal: "=",
+    KeyA: "a",
+    KeyC: "c",
+    KeyD: "d",
+    KeyE: "e",
+    KeyG: "g",
```

Comment:
This map should be made complete, at least for the letters, rather than have only the letters we happen to have shortcuts for today.

That will save people in the future from some confused debugging when they try adding a shortcut in the table of shortcuts and it doesn't seem to do anything (because it isn't also in this table). Confusing especially because they may have done the exact same thing the previous day for a different shortcut and it *did* work (because that one was already here, because there was an existing shortcut with different modifier keys on the same letter).

Reply:
Probably also most readable if the non-letters are grouped together either before or after the letters, rather than be split based on whether their names happen to come before or after "Key" alphabetically.

Reply:
> This map should be made complete, at least for the letters

Done

> Probably also most readable if the non-letters are grouped together

Done